src/flash.rs

```diff
@@ -0,0 +1,165 @@
+use crate::signature::FlashSize;
+use crate::stm32::FLASH;
+use core::{ptr, slice};
+
+/// Flash erase/program error
+#[derive(Debug, Clone, Copy)]
+pub enum Error {
+    ProgrammingSequence,
+    ProgrammingParallelism,
+    ProgrammingAlignment,
+    WriteProtection,
+    Operation,
+}
+
+impl Error {
+    fn read(flash: &FLASH) -> Option<Self> {
+        let sr = flash.sr.read();
+        if sr.pgserr().bit() {
+            Some(Error::ProgrammingSequence)
+        } else if sr.pgperr().bit() {
+            Some(Error::ProgrammingParallelism)
+        } else if sr.pgaerr().bit() {
+            Some(Error::ProgrammingAlignment)
+        } else if sr.wrperr().bit() {
+            Some(Error::WriteProtection)
+        } else if sr.operr().bit() {
+            Some(Error::Operation)
+        } else {
+            None
+        }
+    }
+}
+
+/// Flash methods implemented for `stm32::FLASH`
+pub trait FlashExt {
+    /// Memory-mapped address
+    fn address(&self) -> usize;
+    /// Size in bytes
+    fn len(&self) -> usize;
+    /// Returns a read-only view of flash memory
+    fn read(&self) -> &[u8] {
+        let ptr = self.address() as *const _;
+        unsafe { slice::from_raw_parts(ptr, self.len()) }
+    }
+    /// Unlock flash for erasing/programming until this method's
+    /// result is dropped
+    fn unlocked(&mut self) -> UnlockedFlash;
+}
+
+impl FlashExt for FLASH {
+    fn address(&self) -> usize {
+        0x0800_0000
+    }
+
+    fn len(&self) -> usize {
+        FlashSize::get().bytes()
+    }
+
+    fn unlocked(&mut self) -> UnlockedFlash {
+        unlock(self);
+        UnlockedFlash { flash: self }
+    }
+}
+
+const PSIZE_X8: u8 = 0b00;
+
+/// Result of `FlashExt::unlocked()`
+pub struct UnlockedFlash<'a> {
+    flash: &'a mut FLASH,
+}
+
+/// Automatically lock flash erase/program when leaving scope
+impl Drop for UnlockedFlash<'_> {
+    fn drop(&mut self) {
+        lock(&self.flash);
+    }
+}
+
+impl UnlockedFlash<'_> {
+    /// Erase a flash sector
+    ///
+    /// Refer to the reference manual to see which sector corresponds
+    /// to which memory address.
+    pub fn erase(&mut self, sector: u8) -> Result<(), Error> {
+        let snb = if sector < 12 { sector } else { sector + 4 };
+
+        #[rustfmt::skip]
+        self.flash.cr.modify(|_, w| unsafe {
+            w
+                // start
+                .strt().set_bit()
+                .psize().bits(PSIZE_X8)
+                // sector number
+                .snb().bits(snb)
+                // sectore erase
+                .ser().set_bit()
+                // no programming
+                .pg().clear_bit()
+        });
+        self.wait_ready();
+        self.ok()
+    }
+
+    /// Program bytes with offset into flash memory,
+    /// aligned to 128-bit rows
+    pub fn program<I>(&mut self, mut offset: usize, mut bytes: I) -> Result<(), Error>
+    where
+        I: Iterator<Item = u8>,
```

Comment:
I think this may be more easily used as `IntoIterator`.

Reply:
Isn't `Iterator` more flexible than `IntoInterator` ?

Reply:
No. `IntoIterator` allows you to work with collection types as well as `Iterator` (which automatically implements `IntoIterator`) which is super-handy (but probably not that important in this particular usecase).

Reply:
Oh, just realized that the suggestion was with `Item = &'a u8`

Reply:
It may be best to just pass a `&[u8]`, even. Ultimately, I don't think you gain much by using `IntoIter<Item = &u8>` really. Definitely willing to be convinced (insofar as my opinion matters, hah!).